stdlib/public/runtime/SwiftRT-ELF.cpp

```diff
@@ -0,0 +1,63 @@
+//===--- SwiftRT-ELF.cpp --------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "ImageInspectionELF.h"
+
+#include <cstddef>
+
+// Create empty sections to ensure that the start/stop symbols are synthesized
+// by the linker.  Otherwise, we may end up with undefined symbol references as
+// the linker table section was never constructed.
+
+#define DECLARE_SWIFT_SECTION(name)                                            \
+  __asm__("\t.section " #name ",\"a\"\n");                                     \
+  __attribute__((__visibility__("hidden"))) extern const char __start_##name;  \
+  __attribute__((__visibility__("hidden"))) extern const char __stop_##name;
+
+extern "C" {
+DECLARE_SWIFT_SECTION(swift2_protocol_conformances)
+DECLARE_SWIFT_SECTION(swift2_type_metadata)
+
+DECLARE_SWIFT_SECTION(swift3_typeref)
+DECLARE_SWIFT_SECTION(swift3_reflstr)
+DECLARE_SWIFT_SECTION(swift3_fieldmd)
+DECLARE_SWIFT_SECTION(swift3_assocty)
+}
+
+#undef DECLARE_SWIFT_SECTION
+
+#define SWIFT_SECTION_RANGE(name)                                              \
+  { reinterpret_cast<uintptr_t>(&__start_##name),                              \
+    static_cast<uintptr_t>(&__stop_##name - &__start_##name) }
+
+const swift::MetadataSections sections = {
+    swift::CurrentSectionMetadataVersion,
+    0,
+
+    SWIFT_SECTION_RANGE(swift2_protocol_conformances),
+    SWIFT_SECTION_RANGE(swift2_type_metadata),
+
+    SWIFT_SECTION_RANGE(swift3_typeref),
+    SWIFT_SECTION_RANGE(swift3_reflstr),
+    SWIFT_SECTION_RANGE(swift3_fieldmd),
+    SWIFT_SECTION_RANGE(swift3_assocty),
+};
```

Comment:
Does Clang manage to compile this into a global constant, despite not formally being constexpr by C++ rules?

Reply:
No :-(.  I think that I would rather do that as a follow up.  We can try out a few different things to see if we can do something more clever.

Reply:
SGTM.